l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,720 @@
+
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/access/Ownable2StepUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/security/PausableUpgradeable.sol";
+
+import {IERC20} from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts-v4/token/ERC20/utils/SafeERC20.sol";
+
+import {NEW_ENCODING_VERSION, LEGACY_ENCODING_VERSION} from "./asset-router/IAssetRouterBase.sol";
+import {IL1NativeTokenVault} from "./ntv/IL1NativeTokenVault.sol";
+
+import {IL1ERC20Bridge} from "./interfaces/IL1ERC20Bridge.sol";
+import {IL1AssetRouter} from "./asset-router/IL1AssetRouter.sol";
+import {IAssetRouterBase} from "./asset-router/IAssetRouterBase.sol";
+import {INativeTokenVault} from "./ntv/INativeTokenVault.sol";
+
+import {IL1Nullifier, FinalizeL1DepositParams} from "./interfaces/IL1Nullifier.sol";
+
+import {IMailbox} from "../state-transition/chain-interfaces/IMailbox.sol";
+import {L2Message, TxStatus} from "../common/Messaging.sol";
+import {UnsafeBytes} from "../common/libraries/UnsafeBytes.sol";
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+import {ETH_TOKEN_ADDRESS} from "../common/Config.sol";
+import {DataEncoding} from "../common/libraries/DataEncoding.sol";
+
+import {IBridgehub} from "../bridgehub/IBridgehub.sol";
+import {L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR, L2_ASSET_ROUTER_ADDR} from "../common/L2ContractAddresses.sol";
+import {DataEncoding} from "../common/libraries/DataEncoding.sol";
+import {Unauthorized, SharedBridgeKey, DepositExists, AddressAlreadySet, InvalidProof, DepositDoesNotExist, SharedBridgeValueNotSet, WithdrawalAlreadyFinalized, L2WithdrawalMessageWrongLength, InvalidSelector, SharedBridgeValueNotSet, ZeroAddress} from "../common/L1ContractErrors.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev Bridges assets between L1 and ZK chain, supporting both ETH and ERC20 tokens.
+/// @dev Designed for use with a proxy for upgradability.
+contract L1Nullifier is IL1Nullifier, ReentrancyGuard, Ownable2StepUpgradeable, PausableUpgradeable {
+    using SafeERC20 for IERC20;
+
+    /// @dev Bridgehub smart contract that is used to operate with L2 via asynchronous L2 <-> L1 communication.
+    IBridgehub public immutable override BRIDGE_HUB;
+
+    /// @dev Era's chainID
+    uint256 internal immutable ERA_CHAIN_ID;
+
+    /// @dev The address of ZKsync Era diamond proxy contract.
+    address internal immutable ERA_DIAMOND_PROXY;
+
+    /// @dev Stores the first batch number on the ZKsync Era Diamond Proxy that was settled after Diamond proxy upgrade.
+    /// This variable is used to differentiate between pre-upgrade and post-upgrade Eth withdrawals. Withdrawals from batches older
+    /// than this value are considered to have been finalized prior to the upgrade and handled separately.
+    uint256 internal _eraPostDiamondUpgradeFirstBatch;
```

Comment:
Why `_` prefix on the storage variable? If you want to deprecate the variable you can add prefix `DEPRECATED_` prefix instead

Reply:
Ok, I'll fix